core/src/main/scala/akka/persistence/r2dbc/state/scaladsl/R2dbcDurableStateStore.scala

```diff
@@ -93,22 +116,30 @@ class R2dbcDurableStateStore[A](system: ExtendedActorSystem, config: Config, cfg
       revision,
       DurableStateDao.EmptyDbTimestamp,
       DurableStateDao.EmptyDbTimestamp,
-      serialized,
+      Some(serialized),
       serializer.identifier,
       manifest,
       if (tag.isEmpty) Set.empty else Set(tag))
 
-    stateDao.writeState(serializedRow)
+    stateDao.upsertState(serializedRow)
 
   }
 
   @deprecated(message = "Use the deleteObject overload with revision instead.", since = "1.0.0")
   override def deleteObject(persistenceId: String): Future[Done] =
     deleteObject(persistenceId, revision = 0)
 
+  /**
+   * Delete the value, which will fail with `IllegalStateException` if the existing stored `revision` + 1 isn't equal to
+   * the given `revision`. This optimistic locking check can be disabled with configuration `assert-single-writer`. The
+   * stored revision for the persistenceId is updated and next call to [[getObject]] will return the revision, but with
+   * no value.
+   *
+   * If the given revision is `0` it will fully delete the value and revision from the database without any optimistic
+   * locking check. Next call to [[getObject]] will then return revision 0 and no value.
```

Comment:
Is that how we will give the option of a hard delete? Sound a bit hacky, no?

Wouldn't it be better to have an explicit method for the hard-delete path?

Reply:
We don't have any API for hard delete in Akka DurableStateBehavior, and I don't think we should add it. Resetting revision will break queries/projections. It can be useful for some kind of backoffice tool and then it will be on the store directly. I left the hard delete case mostly for compatibility with previous behavior when using the deprecated deleteObject method without revision parameter. However, I don't think it's completely hacky to use revision=0 because that is what the revision will be after the delete.

Reply:
Ok I see. 

Thanks for the clarifications.